src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h

```diff
@@ -0,0 +1,216 @@
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
+    tmp_buf = MPIDIG_gpu_get_buf(size);
+    yaksa_request_t request;
+    uintptr_t actual_pack_bytes;
+    rc = yaksa_ipack((const void *) buf, (uintptr_t) size, YAKSA_TYPE__BYTE, (uintptr_t) true_lb,
+                     tmp_buf, (uintptr_t) size, &actual_pack_bytes, &request);
+    MPIR_Assert(rc == YAKSA_SUCCESS);
+    rc = yaksa_request_wait(request);
+    MPIR_Assert(rc == YAKSA_SUCCESS);
+
+    tmp_buf -= true_lb;
+
+    MPIDIG_gpu_stage_ctx_t *ctx = MPL_malloc(sizeof(*ctx), MPL_MEM_OTHER);
+    MPIR_Assert(ctx != NULL);
+    ctx->is_send = true;
+    ctx->u.send.true_lb = true_lb;
+    MPIDIU_map_set(MPIDIG_gpu_global.buf_map, (uint64_t) tmp_buf, ctx, MPL_MEM_OTHER);
+#endif
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_unstage_send_buf - free stage buffer */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_unstage_send_buf(const void *buf)
+{
+#ifdef MPL_HAVE_GPU
+    MPIDIG_gpu_stage_ctx_t *ctx =
+        (MPIDIG_gpu_stage_ctx_t *) MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+    MPIDIG_gpu_release_buf((char *) buf + ctx->u.send.true_lb);
+    MPL_free(ctx);
+    MPIDIU_map_erase(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+#endif
+}
+
+/* MPIDIG_gpu_stage_recv_buf - allocate stage buffer and return pointer */
+MPL_STATIC_INLINE_PREFIX void *MPIDIG_gpu_stage_recv_buf(void *buf, MPI_Aint count,
+                                                         MPI_Datatype datatype)
+{
+    char *tmp_buf = (char *) buf;
+#ifdef MPL_HAVE_GPU
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
+    tmp_buf = MPIDIG_gpu_get_buf(size);
+    tmp_buf -= true_lb;
+
+    MPIDIG_gpu_stage_ctx_t *ctx = MPL_malloc(sizeof(*ctx), MPL_MEM_OTHER);
+    MPIR_Assert(ctx != NULL);
+    ctx->is_send = false;
+    ctx->u.recv.dst_buf = buf;
+    ctx->u.recv.count = count;
+    ctx->u.recv.datatype = datatype;
+    MPIDIU_map_set(MPIDIG_gpu_global.buf_map, (uint64_t) tmp_buf, ctx, MPL_MEM_OTHER);
+#endif
+    return tmp_buf;
+}
+
+/* MPIDIG_gpu_unstage_recv_buf - copy data from stage buffer to source buffer and free it */
+MPL_STATIC_INLINE_PREFIX void MPIDIG_gpu_unstage_recv_buf(void *buf)
+{
+#ifdef MPL_HAVE_GPU
+    int rc;
+    MPIDIG_gpu_stage_ctx_t *ctx =
+        (MPIDIG_gpu_stage_ctx_t *) MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf);
+
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+
+    MPIDI_Datatype_get_info(ctx->u.recv.count, ctx->u.recv.datatype, dt_contig, size, dt_ptr,
+                            true_lb);
+    if (!dt_contig)
+        size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
+
```

Comment:
Compilation error:
`/home/caochong/git/mpich-ofi/src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h:180:34: error: ‘count’ undeclared (first use in this function); did you mean ‘conj’?
         size = (dt_ptr->extent * count) + (dt_ptr->true_ub - true_lb) - dt_ptr->extent;
`

Reply:
I think the `count` is get from `ctx` of calling `MPIDIU_map_lookup(MPIDIG_gpu_global.buf_map, (uint64_t) buf)`

Reply:
ctx.count?

Reply:
Yes sorry that should be `ctx->u.recv.count`